homeassistant/components/yeelight/__init__.py

```diff
@@ -139,73 +155,245 @@
+            hass.config_entries.flow.async_init(
+                DOMAIN, context={"source": SOURCE_IMPORT}, data=entry_config,
+            ),
+        )
 
-        _setup_device(hass, config, info[CONF_HOST], device_config)
+    return True
 
-    discovery.listen(hass, SERVICE_YEELIGHT, device_discovered)
 
-    def update(_):
-        for device in list(yeelight_data.values()):
-            device.update()
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Yeelight from a config entry."""
+
+    async def _initialize(is_discovery) -> None:
+        unset = entry.add_update_listener(_async_update_listener)
+        hass.data[DOMAIN][DATA_CONFIG_ENTRIES][entry.entry_id] = {
+            DATA_UNSUB_UPDATE_LISTENER: unset,
+        }
+        if is_discovery:
+            scanner = YeelightScanner(hass, entry)
+            hass.data[DOMAIN][DATA_CONFIG_ENTRIES][entry.entry_id][
+                DATA_SCANNER
+            ] = scanner
+
+        platform_setups = [
+            hass.config_entries.async_forward_entry_setup(entry, component)
+            for component in PLATFORMS
+        ]
+
+        if is_discovery:
+            # Wait for platform setup is important. Make sure callbacks are set up before scanning.
+            await asyncio.gather(*platform_setups)
+            await scanner.start_scan()
+        else:
+            for platform_setup in platform_setups:
+                hass.async_create_task(platform_setup)
+
+    if entry.data[CONF_DISCOVERY]:
+        hass.async_create_task(_initialize(True))
+    else:
+
+        # Move options from data for imported entries
+        if not entry.options:
+            hass.config_entries.async_update_entry(
+                entry,
+                data={
+                    CONF_DISCOVERY: entry.data.get(CONF_DISCOVERY),
+                    CONF_NAME: entry.data.get(CONF_NAME),
+                    CONF_IP_ADDRESS: entry.data[CONF_IP_ADDRESS],
+                },
+                options={
+                    CONF_MODEL: entry.data.get(CONF_MODEL, ""),
+                    CONF_TRANSITION: entry.data.get(
+                        CONF_TRANSITION, DEFAULT_TRANSITION
+                    ),
+                    CONF_MODE_MUSIC: entry.data.get(
+                        CONF_MODE_MUSIC, DEFAULT_MODE_MUSIC
+                    ),
+                    CONF_SAVE_ON_CHANGE: entry.data.get(
+                        CONF_SAVE_ON_CHANGE, DEFAULT_SAVE_ON_CHANGE
+                    ),
+                    CONF_NIGHTLIGHT_SWITCH: entry.data.get(
+                        CONF_NIGHTLIGHT_SWITCH, DEFAULT_NIGHTLIGHT_SWITCH
+                    ),
+                },
+            )
+
+        config = {
+            CONF_NAME: entry.data.get(CONF_NAME),  # Support name import from yaml
+            **entry.options,
+        }
+        await _async_setup_device(hass, entry.data[CONF_IP_ADDRESS], config)
+
+        await _initialize(False)
```

Comment:
Can we drop the variation for the platform set up and always create a task for initialization regardless of discovery?